### ISSUE TITLE
aws_network_interface data source and resource provide different information

### ISSUE THREAD

**Issue by gwvandesteeg** [terraform-providers/terraform-provider-aws#6863]
The aws_network_interface data source and related resource both provide different information back.

Documentation issue:
* resource aws_network_interface, returned values doesn't list "id" as one of the returned values

Code and Documentation issue on the resource:
* The values returned from the AWS CLI command for create-network-interface and describe-network-interface (along with the relevant API calls) return the MAC address and many more attributes which are found in the data source use of network_interface.  These should also be available in the resulting creation of the network interface.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.9
terraform-provider-aws 1.51

### Affected Resource(s)

* aws_network_interface

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp

resource "aws_vpc" "test" {
  cidr_block = "10.0.0.0/16"
}

resource "aws_subnet" "az" {
  vpc_id        = "${aws_vpc.test.id}"
  cidr_block  = "10.0.1.0/24"
}

resource "aws_network_interface" "net0" {
  subnet_id = "${aws_subnet.az.id}"
  private_ips = ["10.0.1.10"]
}

data "aws_network_interface" "nic0" {
  id = "${aws_network_interface.net0.id}"
}

output "data_mac" {
  value = "${data.aws_network_interface.nic0.mac_address}"
}

output "resource_mac" {
  value = "${aws_network_interface.net0.mac_address}''
}
```

### Expected Behavior

Output of the MAC address
```
Outputs:

data_mac = 00:11:22:33:44:55
resource_mac = 00:11:22:33:44:55
```
### Actual Behavior

```
Error: Error running plan: 1 error(s) occurred:

* output.resource_mac: Resource 'aws_network_interface.net0' does not have attribute 'mac_address'

```
### Steps to Reproduce

1. `terraform apply`

### References

* https://docs.aws.amazon.com/cli/latest/reference/ec2/create-network-interface.html
* https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateNetworkInterface.html
* #6125 - Related for documentation issue

**Comment by henrystrick** [comment]
Looks like an interesting issue to me. And important if it would reflect a bug. 
There are several things to ask though, immediately.

Did you run terraform plan with this test input? What did the file say about what happened? That  would probably give you a clue, if not the solution.

Also, what were you going to do with this interface. My experience with terraform for aws, though not extensive, is that if you set up a complete EC2 "environment", i.e. your own VPC, with subnets, and one or two instances, terraform in AWS automatically creates network interfaces for your instance(s), even if you do not specify it in your terraform config files. And this interface is not a requester-managed, but a proper new interface controlled / managed by myself.

Now the behaviour that I describe is perhaps a bit strange (as they were created without me explicitly asking for them) - but it might have something to do with what you observe, or what you describe in the above data. 

I just wonder because you only ran aws_network_interface , without attaching it, perhaps it was not (yet) fully created?! Probably terraform plan should tell you more about this.

Lastly, did you at least run
aws ec2 describe-network-interfaces for the whole region, or at least for the test vpc you created? That might also produce helpful information.

Hope this helps. If you does, and the situation is clearer, or if you have already resolved it yourself, could you let us know here? That would be of general interest. Many thanks!

**Comment by gwvandesteeg** [comment]
It is quite simple actually,

Terraform plan and attaching the resource to an instance are quite
irrelevant in this situation, the configuration above shows the absolute
minimum needed to reproduce the difference, you have to remember that a
network interface created can be attached to more things than an EC2
instance.  But when dealing with an EC2 instance it is definitely important
to get the right information.

Now, because the terraform resource does not return the same information as
the terraform data source (even though the API and CLI commands do) this
means that when you create the interface you cannot then pass this
information such as the MAC address, the IPv6 addresses, the primary
private IPv4 (you get a list of ALL IPv4s back on the resource, not the
primary private IPv4 and the list of secondary private IPv4s as you do on
the data source) on to any further resources you wish to create.  Such as
passing the dynamically allocated IP addresses through to creating custom
Route53 DNS entries (for both IPv4 or IPv6) or to AWS Cloud Map for service
discovery, or for passing the MAC address (or other details of the
interface) through to an EC2 instance via the cloud-init configuration that
is being created for various purposes.  The latter of which is especially
important when creating EC2 instances with multiple network interfaces.
Doubly so when you need them to be attached in the correct order (which is
where you don't want to use the dynamically allocated interface on an EC2
instance, but explicitly create all attached interfaces) because you're
using multiple subnets in your VPC.

**Comment by bflad** [comment]
Support for the `mac_address` attribute in the `aws_network_interface` resource has been merged and will release with version 2.46.0 of the Terraform AWS Provider, next Thursday. Thanks to @khalednassar for the implementation. 👍 If you have further bug reports or feature requests, please submit new GitHub issues following the templates and we will take a fresh look.
